sklearn/metrics/pairwise.py

```diff
@@ -557,9 +607,54 @@ def cosine_distances(X, Y=None):
     S = cosine_similarity(X, Y)
     S *= -1
     S += 1
+
+    # Ensure non negativity.
+    # This may not be the case due to floating point rounding errors.
+    np.maximum(S, 0, out=S)
+    if X is Y:
+        # Ensure that distances between vectors and themselves are set to 0.0.
+        # This may not be the case due to floating point rounding errors.
+        S.flat[::S.shape[0] + 1] = 0.0
+
     return S
 
 
+def _validate_binary_pairwise_input(X, Y):
+    """Check for constant input. Those have to be handled separately, as scipy
+    would return nans.
+
+    np.ptr is more robust than np.mean for checking if a vector is constant.
+
+    For example:
+    >>> import numpy as np
+    >>> x = [0.667, 0.667, 0.667]
+    >>> x_center = x - np.mean(x)  # it is negative due to numerical issues
+    >>> x_center < 0
+    array([ True,  True,  True], dtype=bool)
+    >>> np.ptp(x) == 0
+    True
+    """
+    const_X_idx = np.where(np.ptp(X, axis=1) == 0)[0]
+
+    if X is Y:
+        const_Y_idx = const_X_idx
+    else:
+        const_Y_idx = np.where(np.ptp(Y, axis=1) == 0)[0]
+
+    return const_X_idx, const_Y_idx
```

Comment:
I'd just put all this code in ``_filter_binary_pairwise_output``. No need for the indirection to another function, since it's so short and only used once.